src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,698 @@
+        {
+            List<KeyValuePair<string, object>> uniqueHoles = null;
+            for (int i = 0; i < value.Count; ++i)
+            {
+                for (int j = i + 1; j < value.Count; ++j)
+                {
+                    if (string.Equals(value[i].Key, value[j].Key, StringComparison.Ordinal))
+                    {
+                        uniqueHoles = new List<KeyValuePair<string, object>>();
+                        for (int k = 0; k < value.Count; ++k)
+                        {
+                            if (k <= i)
+                            {
+                                uniqueHoles.Add(value[k]);
+                            }
+                            else
+                            {
+                                bool duplicateKey = false;
+                                foreach (var hole in uniqueHoles)
+                                {
+                                    if (string.Equals(hole.Key, value[k].Key, StringComparison.Ordinal))
+                                    {
+                                        duplicateKey = true;
+                                        break;
+                                    }
+                                }
+                                if (!duplicateKey)
+                                    uniqueHoles.Add(value[k]);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return uniqueHoles ?? value;
+        }
+
+        /// <inheritDoc/>
+        public object this[object key]
+        {
+            get
+            {
+                object value;
+                if (_eventProperties != null && _eventProperties.TryGetValue(key, out value))
+                    return value;
+
+                int keyIndex = MessagePropertiesIndexOfKey(key);
+                if (keyIndex != -1)
+                {
+                    return _messageProperties[keyIndex].Value;
+                }
+
+                throw new KeyNotFoundException();
+            }
+            set
+            {
+                EventProperties[key] = value;
+            }
+        }
+
+        /// <inheritDoc/>
+        public ICollection<object> Keys { get { return (ICollection<object>)KeyCollection ?? new DictionaryCollection(this, true); } }
+
+        /// <inheritDoc/>
+        public ICollection<object> Values { get { return (ICollection<object>)ValueCollection ?? new DictionaryCollection(this, false); } }
+
+        private Dictionary<object, object>.KeyCollection KeyCollection
+        {
+            get
+            {
+                if (_eventProperties != null && (_messageProperties == null || _messageProperties.Count == 0))
+                    return _eventProperties.Keys;
+                if (_eventProperties == null && (_messageProperties == null || _messageProperties.Count == 0))
```

Comment:
I prefer an `else` and nested `if`s here. 

Maybe also a helper for `_messageProperties == null || _messageProperties.Count == 0` (9 times used?)

Reply:
Fixed